### PR TITLE
feat: add Particle.is_lepton method

### PR DESCRIPTION
**Pull request by redeboer**
While the title of this PR was the original motivation (it is implemented in #173), the bulk of this PR is actually the reorganisation of the tests for the `data` module. This allowed to spot some mistakes and clean up duplicate tests. The convention followed there is explained in a new section in the contribute page.

Some other improvements:
* The test of `Particle.__repr__` didn't do anything useful, but that has been fixed. This is relevant for the discussion in #260.
* Test collection warnings are treated now treated as errors, because there's a risk certain tests you implemented are not run.
* For VSCode: the `__init__.py` files won't be shown in search or when using code navigation (e.g. `Ctrl+P`).
* Test coverage for the `data` module has improved slightly.

### CODE REVIEW

File: tests/unit/test_data.py [hunk 1]
```diff
@@ -75,8 +75,8 @@ def test_exceptions(magnitude, projection):
 
 class TestParticle:
     @staticmethod
-    def test_repr(particle_database):
-        for particle in particle_database:
+    def test_repr(particle_database: ParticleCollection):
+        for particle in particle_database.values():
```
Comment (spflueger) [inline tests/unit/test_data.py:100]: Just a comment, the `values()` seems kind of redundant. I guess the `ParticleCollection` interface is `add`, `find` and iteration. Something for another PR though

File: expertsystem/data.py [hunk 2]
```diff
@@ -296,10 +296,14 @@ class ParticleCollection(abc.Mapping):
...
-            if isinstance(particles, (list, set, tuple)):
-                self.__particles.update(
-                    {particle.name: particle for particle in particles}
+            if not isinstance(particles, (list, set, tuple)):
+                raise ValueError(
+                    f"Cannot construct a {self.__class__.__name__} "
+                    f"from a {particles.__class__.__name__}"
                 )
+            self.__particles.update(
+                {particle.name: particle for particle in particles}
```
Comment (spflueger) [inline expertsystem/data.py:305]: Improvement suggestion: Actually there is no runtime check on what `particle` is. So theoretically any kind of object can be injected here as long as it has a attribute `name`.

Comment (redeboer) [reply expertsystem/data.py]: 26fae1f1
I don't want to slow down construction too much though, that's why I put it in the comprehension

File: tests/unit/test_data.py (hunk 1 above)
Comment (redeboer) [reply tests/unit/test_data.py:100]: The reason is that the `ParticleCollection` mimics an (immutable) `Mapping`:
https://github.com/ComPWA/expertsystem/blob/26fae1f1b407ad870ec227635e226fa19d1e9a77/expertsystem/data.py#L293

**Review by spflueger — APPROVED** [review]

**Pull request merged**